dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/WorkflowSpecifications.java

```diff
@@ -0,0 +1,34 @@
+package io.dockstore.webservice.helpers;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Contains information used to create/update a Workflow/WorkflowVersion
+ * that is not necessarily part of those objects themselves.  An instance
+ * of this class is intended as an ephemeral vessel used to pass
+ * information between the code that parses user input (.dockstore.yml)
+ * and the workflow/version creation code.
+ *
+ * This class relates to workflows/versions as a blueprint does to a
+ * building, or a recipe to dinner: it contains information used to create
+ * the object, but it itself is not part of the object.
+ */
+public class WorkflowSpecifications {
```

Comment:
I think the other part that isn't obvious to me yet, if this is ephemeral and is used to guide what files should be indexed on creation, what happens when the notebook/workflow is updated? How do we know what files to index on an update, by looking at a previous version?

Reply:
Similar to the other information like "authors" in `.dockstore.yml`, when a new branch is pushed, we retrieve the new list of files (currently from .dockstore.yml) and stick it in an instance of `WorkflowSpecifications`, which fans out to the code that needs it.  The user-specified list of files can vary (and probably will vary) from version to version.

Reply:
Ok, I see. 
So `it contains information used to create the object, but it itself is not part of the object` is kind of like `it contains information used to create and update the object`

Reply:
Yeah, I'm going to elaborate. I think I understand what this is:

(my understanding) a source of information about workflows that is neither saved to the DB or frozen along with snapshots. 

What I don't understand:

* how a developer would decide to put stuff in here or not going forward. For example, `VersionMetadata` clearly stores stuff that we want to keep long-term like DOIs or ORCID authors but should not be subject to snapshots. The stuff in the `WorkflowVersion` is the other information that can be frozen and should not be tampered with as an immutable reproducible workflow. 

This has a path like `WorkflowVersion`, are these paths different in some way that they should be able to be changed by the user even after a snapshot? Is this supposed to be associated with a `Workflow` or `WorkflowVersion`? I would have thought that a list of extra files would be associated with the `WorkflowVersion` but the javadoc is ambiguous. 
Is there a reason why this should not be frozen? etc.

Reply:
> Yeah, I'm going to elaborate. I think I understand what this is:
> 
> (my understanding) a source of information about workflows that is neither saved to the DB or frozen along with snapshots.
> 
> What I don't understand:
> 
> * how a developer would decide to put stuff in here or not going forward. For example, `VersionMetadata` clearly stores stuff that we want to keep long-term like DOIs or ORCID authors but should not be subject to snapshots. The stuff in the `WorkflowVersion` is the other information that can be frozen and should not be tampered with as an immutable reproducible workflow.
> 
> This has a path like `WorkflowVersion`, are these paths different in some way that they should be able to be changed by the user even after a snapshot? Is this supposed to be associated with a `Workflow` or `WorkflowVersion`? I would have thought that a list of extra files would be associated with the `WorkflowVersion` but the javadoc is ambiguous. Is there a reason why this should not be frozen? etc.

`WorkflowSpecifications` is like a bag we put information about how to create/update the workflow/version into so we can carry it to the parts of the code that need it.  Said information can be saved (perhaps literally or in a translated form) in the appropriate objects and to the DB as needed.  As currently, in the PR, there's a list of user-specified paths which is passed to the language handler, which it uses to read files from the repo and create a set of corresponding `SourceFiles`, which are then added to a `WorkflowVersion`.  The `WorkflowVersion` can be frozen/unfrozen as usual.

So, while we're not persisting the list of paths, we are persisting the resulting SourceFiles.  There's nothing stopping us from persisting the list of paths itself, but we don't persist the list for Services, and it seemed like redundant information.  Since it comes from `.dockstore.yml`, a new [and possibly different] list will be present with each push.

Reply:
> there's a list of user-specified paths which is passed to the language handler, which it uses to read files from the repo and create a set of corresponding `SourceFiles`, which are then added to a `WorkflowVersion`

I get that, the question is more why aren't these user-specified paths added to the version like they are with `Service.workflowpath` Is there some distinction drawn here why this wasn't put in the metadata or the object?